subsys/net/lib/nrf_cloud/src/nrf_cloud_jwt.c

```diff
@@ -8,12 +8,451 @@
 #include <zephyr/kernel.h>
 #include <zephyr/logging/log.h>
 #include <modem/modem_jwt.h>
+#if defined(CONFIG_MODEM_JWT)
 #include <nrf_modem_at.h>
+#endif
 
 #define GET_TIME_CMD "AT%%CCLK?"
 
 LOG_MODULE_REGISTER(nrf_cloud_jwt, CONFIG_NRF_CLOUD_LOG_LEVEL);
 
+#if defined(CONFIG_NRF_CLOUD_JWT_SOURCE_CUSTOM)
+#include <zephyr/sys/base64.h>
+#include <zephyr/random/random.h>
+
+#include <tinycrypt/ctr_prng.h>
+#include <tinycrypt/sha256.h>
+#include <tinycrypt/ecc_dsa.h>
+#include <tinycrypt/constants.h>
+
+#include <date_time.h>
+#include <cJSON.h>
+
+#include "nrf_cloud_mem.h"
+
+#define ECDSA_SHA_256_SIG_SZ	(64)
+#define ECDSA_SHA_256_HASH_SZ	(32)
+#define BASE64_ENCODE_SZ(n)	(((4 * n / 3) + 3) & ~3)
+#define B64_SIG_SZ		(BASE64_ENCODE_SZ(ECDSA_SHA_256_SIG_SZ) + 1)
+
+/* JWT header for ES256 type: {"typ":"JWT","alg":"ES256"} */
+#define JWT_ES256_HDR		"eyJ0eXAiOiJKV1QiLCJhbGciOiJFUzI1NiJ9"
+
+/* Size of the ES256 private key material */
+#define PRV_KEY_SZ		(32)
+
+#define PRV_KEY_DER_SZ		(138)
+#define PRV_KEY_PEM_SZ		(256)
+#define PRV_KEY_DER_START_IDX	(36)
+
+static TCCtrPrng_t prng_state;
+
+static int setup_prng(void)
+{
+	static bool prng_init;
+	uint8_t entropy[TC_AES_KEY_SIZE + TC_AES_BLOCK_SIZE];
+
+	if (prng_init) {
+		return 0;
+	}
+
+	for (int i = 0; i < sizeof(entropy); i += sizeof(uint32_t)) {
+		uint32_t rv = sys_rand32_get();
+
+		memcpy(entropy + i, &rv, sizeof(uint32_t));
+	}
+
+	int res = tc_ctr_prng_init(&prng_state,
+				   (const uint8_t *)&entropy, sizeof(entropy),
+				   __FILE__, sizeof(__FILE__));
+
+	if (res != TC_CRYPTO_SUCCESS) {
+		LOG_ERR("tc_ctr_prng_init() failed, error: %d", res);
+		return -EINVAL;
+	}
+
+	prng_init = true;
+
+	return 0;
+}
+
+int default_CSPRNG(uint8_t *dest, unsigned int size)
+{
+	return tc_ctr_prng_generate(&prng_state, NULL, 0, dest, size);
+}
+
+static void base64_url_format(char *const base64_string)
+{
+	if (base64_string == NULL) {
+		return;
+	}
+
+	char *found = NULL;
+
+	/* replace '+' with "-" */
+	for (found = base64_string; (found = strchr(found, '+'));) {
+		*found = '-';
+	}
+
+	/* replace '/' with "_" */
+	for (found = base64_string; (found = strchr(found, '/'));) {
+		*found = '_';
+	}
+
+	/* remove padding '=' */
+	found = strchr(base64_string, '=');
+	if (found) {
+		*found = '\0';
+	}
+}
+
+static int sign_message_der(const char *const der_key, const char *const msg_in,
+			    char *const sig_out)
+{
+	struct tc_sha256_state_struct ctx;
+	uint8_t hash[ECDSA_SHA_256_HASH_SZ];
+	int res;
+
+	tc_sha256_init(&ctx);
+	tc_sha256_update(&ctx, msg_in, strlen(msg_in));
+	tc_sha256_final(hash, &ctx);
```

Comment:
The one thing I wonder is whether we should perhaps use mbedtls and psa crypto functions instead of tinycrypt, since we are targeting mbedtls on non-modem builds anyways

In the short term, this would save you having to manually parse PEM/DER. In the long term, this would save you having to refactor this code (as much) when we switch to opaque keys

It's not 100% clear to me how opaque keys will be handled in this case. But I know it will need to involve using mbedtls/psa crypto, so switching to that now will knock one thing off the list.

Reply:
this PR is already a refactor of my [original effort](https://github.com/nrfconnect/sdk-nrf/compare/main...jayteemo:fw-nrfconnect-nrf:IRIS-5723_nrf_jwt).  
this is just to get something working until PSA is available.
i view this as a temporary/experimental solution.
i'll let @plskeggs decide how much more work needs to be put into this.

Reply:
Let's keep it as is and revisit when switching to opaque.